posts/2018-08-25-http4s-error-handling-mtl.md

```diff
@@ -0,0 +1,333 @@
+
+case class CustomError(msg: String) extends Throwable
+
+def customHandle[F[_], A](f: F[A], fallback: F[A])(implicit ev: MonadError[F, CustomError]): F[A] =
+  f.handleErrorWith(_ => fallback)
+
+val io: IO[Int] = IO(Random.nextInt(2)).flatMap { case 1 => IO.raiseError(new Exception("boom")) }
+customHandle(io, IO.pure(123))
+```
+
+#### Generalizing Http Error Handling
+
+Now back to our use case. We can't have a `MonadError[F, UserError]` constraint because there's already a `MonadError[F, Throwable]` in scope given our `Sync[F]` constraint. But it turns out we can make this work if we also abstract over the error handling by introducing an `HttpErrorHandler` algebra where the error type is a subtype of `Throwable`.
+
+```tut:book:silent
+trait HttpErrorHandler[F[_], E <: Throwable] {
+  def handle(routes: HttpRoutes[F]): HttpRoutes[F]
+}
+
+object HttpErrorHandler {
+  def apply[F[_], E <: Throwable](implicit ev: HttpErrorHandler[F, E]) = ev
+}
+
+object syntax {
+  implicit class HttpErrorHandlerOps[F[_], E <: Throwable](routes: HttpRoutes[F]) {
+    def handleHttpErrorResponse(implicit H: HttpErrorHandler[F, E]): HttpRoutes[F] =
+      H.handle(routes)
+  }
+}
+```
+
+`UserRoutes` can now have an additional constraint of type `HttpErrorHandler[F, UserError]` so we clearly know what kind of errors we are dealing with and can have the Scala compiler on our side.
+
+```tut:book:silent
+class UserRoutesMTL[F[_]: Sync](userAlgebra: UserAlgebra[F])(implicit ev: HttpErrorHandler[F, UserError]) extends Http4sDsl[F] {
+
+  import syntax._
+
+  private val httpRoutes: HttpRoutes[F] = HttpRoutes.of[F] {
+
+    case GET -> Root / "users" / username =>
+      userAlgebra.find(username).flatMap {
+        case Some(user) => Ok(user.asJson)
+        case None => NotFound(username.asJson)
+      }
+
+    case req @ POST -> Root / "users" =>
+      req.as[User].flatMap { user =>
+        userAlgebra.save(user) *> Created(user.username.asJson)
+      }
+
+    case req @ PUT -> Root / "users" / username =>
+      req.as[UserUpdateAge].flatMap { userUpdate =>
+        userAlgebra.updateAge(username, userUpdate.age) *> Created(username.asJson)
+      }
+  }
+
+  val routes: HttpRoutes[F] = httpRoutes.handleHttpErrorResponse
+
+}
+```
+
+We are basically delegating the error handling (AKA mapping business errors to appropiate http responses) to a specific algebra by making use of the syntax we have previously introduced.
+
+We also need an implementation for this algebra in order to handle errors of type `UserError` but first we can introduce a `RoutesHttpErrorHandler` object that encapsulates the repetitive task of handling errors given an `HttpRoutes[F]` to generalize things a bit more :)
+
+```tut:book:silent
+import cats.data.{Kleisli, OptionT}
+
+object RoutesHttpErrorHandler {
+  def apply[F[_], E <: Throwable](routes: HttpRoutes[F])(handler: E => F[Response[F]])(implicit M: MonadError[F, E]): HttpRoutes[F] =
+    Kleisli { req: Request[F] =>
```

Comment:
I can't get `tut` to like this, any compiler flag that needs to be set?

Reply:
I don't see that `-Ypartial-unification` is on here, which often helps with inference puzzlers.  Though setting it here in the SBT repl didn't seem to make a difference.  It works if you're very explicit about types, and we don't have kind-projector in this blog environment, which makes it worse:

```scala
    Kleisli[({ type T[A] = OptionT[F, A] })#T, Request[F], Response[F]] { req: Request[F] =>
      OptionT {
        routes.run(req).value.handleErrorWith { e => handler(e).map(Option(_)) }
      }
    }
```

There's also an `HttpRoutes.apply`, but that takes a `Sync` constraint, whose `MonadError` fights with `M`:

```scala
  def apply2[F[_], E <: Throwable](routes: HttpRoutes[F])(handler: E => F[Response[F]])(implicit F: Sync[F], M: MonadError[F, E]): HttpRoutes[F] =
    HttpRoutes { req: Request[F] =>
      OptionT {
        M.handleErrorWith(routes.run(req).value) { e => M.map(handler(e))(Option(_)) }
      }
    }
```

Reply:
Thanks for investigating Ross.

I guess that if I have to sacrifice readability I'd rather not `tut`... Surely that was a question I wanted to ask, `kind-projector` is not supported then?

Reply:
All of this can be implemented (both the flags and the compiler plugin). It's just that somebody's got to do it [here](https://github.com/typelevel/typelevel.github.com/tree/development/src/main/scala).

Reply:
@larsrh interesting. If you or somebody else can provide any guidance I can do it as part of this PR. For example how do I pass `scalacOptions` settings for `tut`?